metaflow/plugins/cards/component_serializer.py

```diff
@@ -16,11 +23,327 @@ def get_card_class(card_type):
+        user_set_card_id=None,
+        runtime_card=False,
+        card_options=None,
+    ):
+        self._card_creator_args = dict(
+            card_uuid=card_uuid,
+            user_set_card_id=user_set_card_id,
+            runtime_card=runtime_card,
+            decorator_attributes=decorator_attributes,
+            card_options=card_options,
+            logger=logger,
+        )
+        self._card_creator = card_creator
+        self._last_layout_change = None
+        self._latest_user_data = None
+        self._last_refresh = 0
+        self._last_render = 0
+        self._render_seq = 0
+        self._logger = logger
+        self._no_warnings = no_warnings
+        self._warn_once = {
+            "update": {},
+            "not_implemented": {},
+        }
+        card_type = decorator_attributes["type"]
+
+        if components is None:
+            self._components = ComponentStore(
+                logger=self._logger,
+                card_type=card_type,
+                user_set_id=user_set_card_id,
+                components=None,
+            )
+        else:
+            self._components = ComponentStore(
+                logger=self._logger,
+                card_type=card_type,
+                user_set_id=user_set_card_id,
+                components=list(components),
+            )
+
+    def append(self, component, id=None):
+        self._components.append(component, id=id)
+
+    def extend(self, components):
+        self._components.extend(components)
+
+    def clear(self):
+        self._components.clear()
+
+    def _card_proc(self, mode):
+        self._card_creator.create(**self._card_creator_args, mode=mode)
+
+    def refresh(self, data=None, force=False):
+        self._latest_user_data = data
+        nu = time.time()
+
+        # We set this so that when the self._last_layout_change is not the same
+        # as `self.components.layout_last_changed_on`, then the component array itself
+        # has been modified. So we should force a re-render of the card.
+        if self._last_layout_change is None:
+            self._last_layout_change = self.components.layout_last_changed_on
+        # FIXME: make `RUNTIME_CARD_MIN_REFRESH_INTERVAL` customizable at decorator level.
+        if nu - self._last_refresh < RUNTIME_CARD_MIN_REFRESH_INTERVAL:
+            # rate limit refreshes: silently ignore requests that
+            # happen too frequently
+            return
+        self._last_refresh = nu
+
+        # This block of code will render the card in `render_runtime` mode when:
+        # 1. refresh is called with `force=True`
+        # 2. Layout of the components in the card has changed. i.e. The actual elements in the component array have changed.
+        # 3. The last time the card was rendered was more the minimum interval after which they should be rendered.
+        last_rendered_before_minimum_interval = (
+            nu - self._last_refresh
+        ) > RUNTIME_CARD_RENDER_INTERVAL
```

Comment:
@tuulos : Where should we provide users to set `RUNTIME_CARD_RENDER_INTERVAL` and `RUNTIME_CARD_MIN_REFRESH_INTERVAL`; Should we add it to : 
1. metaflow_config.py ?
2. `@card` parameters (what should be the name ?)
3. `MetaflowCard` (should we create a new attribute ?)

One recommendation : 
- Make `RUNTIME_CARD_RENDER_INTERVAL` configurable via `MetaflowCard` and `@card` decorator since this is card-specific. The key idea is that the developer of a card can provide sensible defaults for this value and Metaflow users can over-ride these values. Since this variable governs how often `card create` calls will happen, the best objects to provide such information are the card itself, and the user setting the @card decorator.  
- Make `RUNTIME_CARD_MIN_REFRESH_INTERVAL` a metaflow_config variable. It governs when how often we hit the logic that makes the decision of calling `render_runtime` or `refresh` and behaves like a rate-limiter . _It would also make sense in the MetaflowCard class._

Reply:
we can easily start simply and add more knobs later. We could add `refresh_interval` to `@card` as a starting point. It shouldn't be too hard, right?